node/core/provisioner/src/lib.rs

```diff
@@ -126,70 +168,90 @@ impl ProvisioningJob {
+					Some(RequestInherentData(_, return_sender)) => {
+						let _timer = self.metrics.time_request_inherent_data();
+
+						if self.inherent_after.is_ready() {
+							self.send_inherent_data(vec![return_sender]).await;
+						} else {
+							self.awaiting_inherent.push(return_sender);
 						}
+
 					}
-					self.note_provisionable_data(data);
-
-					// clean up our list of channels by removing the bad indices
-					// start by reversing it for efficient pop
-					bad_indices.reverse();
-					// Vec::retain would be nicer here, but it doesn't provide
-					// an easy API for retaining by index, so we re-collect instead.
-					self.provisionable_data_channels = self
-						.provisionable_data_channels
-						.into_iter()
-						.enumerate()
-						.filter(|(idx, _)| {
-							if bad_indices.is_empty() {
-								return true;
-							}
-							let tail = bad_indices[bad_indices.len() - 1];
-							let retain = *idx != tail;
-							if *idx >= tail {
-								let _ = bad_indices.pop();
+					Some(RequestBlockAuthorshipData(_, sender)) => {
+						self.provisionable_data_channels.push(sender)
+					}
+					Some(ProvisionableData(_, data)) => {
+						let _timer = self.metrics.time_provisionable_data();
+
+						let mut bad_indices = Vec::new();
+						for (idx, channel) in self.provisionable_data_channels.iter_mut().enumerate() {
+							match channel.send(data.clone()).await {
+								Ok(_) => {}
+								Err(_) => bad_indices.push(idx),
 							}
-							retain
-						})
-						.map(|(_, item)| item)
-						.collect();
+						}
+						self.note_provisionable_data(data);
+
+						// clean up our list of channels by removing the bad indices
+						// start by reversing it for efficient pop
+						bad_indices.reverse();
+						// Vec::retain would be nicer here, but it doesn't provide
+						// an easy API for retaining by index, so we re-collect instead.
+						self.provisionable_data_channels = self
+							.provisionable_data_channels
+							.into_iter()
+							.enumerate()
+							.filter(|(idx, _)| {
+								if bad_indices.is_empty() {
+									return true;
+								}
+								let tail = bad_indices[bad_indices.len() - 1];
+								let retain = *idx != tail;
+								if *idx >= tail {
+									let _ = bad_indices.pop();
+								}
+								retain
+							})
+							.map(|(_, item)| item)
+							.collect();
+					}
+					None => break,
+				},
+				_ = self.inherent_after.ready().fuse() => {
+					let return_senders = std::mem::replace(&mut self.awaiting_inherent, Vec::new());
+					if !return_senders.is_empty() {
+						self.send_inherent_data(return_senders).await;
+					}
 				}
-				None => break,
 			}
 		}
 
 		Ok(())
 	}
 
+	async fn send_inherent_data(
+		&mut self,
+		return_senders: Vec<oneshot::Sender<ProvisionerInherentData>>,
+	) {
+		if let Err(err) = send_inherent_data(
+			self.relay_parent,
+			&self.signed_bitfields,
+			&self.backed_candidates,
+			return_senders,
+			self.sender.clone(),
```

Comment:
```suggestion
			&mut self.sender,
```

Reply:
This interferes with the `Send` implementation - i tried this locally and got compilation errors